### PR TITLE
Treat empty array and nil as equivalent when diffing

### PR DESCRIPTION
**Pull request by barney-s**
Treat an empty array and nil as equivalent 

Fix for #289

### CODE REVIEW

**Comment by barney-s** [comment]
@Manny2014 FYI

**Comment by barney-s** [comment]
Tested by porting these changes over to e2e PR.
`    testcase.go:193: Running step3: Update RGD to set the deployment spec to have an empty env array`


```
❯ make test-e2e
go test -v ./test/e2e/suites/basic/...
=== RUN   TestAutoscaledDeploymentRGD
    multi_resource_test.go:38: Using namespace: e2e-autoscaled-deployment-rgd-1739345557, from: testdata/../../../../test/e2e/testdata/autoscaled-deployment-rgd
    testcase.go:193: Running step0: Install the RGD AutoscaledDeployment that creates a ConfigMap, Service, Deployment, and HPA 
    testcase.go:359:    Applying ResourceGraphDefinition/autoscaled-deployment-rgd[rgd.yaml]
    verify.go:57:       Verifying ResourceGraphDefinition/autoscaled-deployment-rgd[0rgd.yaml]
    verify.go:57:       Verifying CustomResourceDefinition/autoscaleddeployments.kro.run[1instancecrd.yaml]
    testcase.go:193: Running step1: Create an instance of AutoscaledDeployment with initial settings (version 1.24, 1 replica, 20% CPU utilization) 
    testcase.go:359:    Applying AutoscaledDeployment/test-app-instance[instance.yaml]
    verify.go:57:       Verifying Deployment/test-app-deployment[deployment.yaml]
    verify.go:57:       Verifying HorizontalPodAutoscaler/test-app-hpa[hpa.yaml]
    verify.go:57:       Verifying AutoscaledDeployment/test-app-instance[instance.yaml]
    verify.go:57:       Verifying Service/test-app-service[service.yaml]
    testcase.go:193: Running step2: Update the instance to use a newer version (1.25) which should trigger a rolling update of the Deployment 
    testcase.go:359:    Applying AutoscaledDeployment/test-app-instance[instance.yaml]
    verify.go:57:       Verifying Deployment/test-app-deployment[deployment.yaml]
    testcase.go:193: Running step3: Update the instance to increase replicas to 3 and adjust HPA settings (min: 2, max: 5) 
    testcase.go:359:    Applying AutoscaledDeployment/test-app-instance[instance.yaml]
    verify.go:57:       Verifying Deployment/test-app-deployment[deployment.yaml]
    verify.go:57:       Verifying HorizontalPodAutoscaler/test-app-hpa[hpa.yaml]
    testcase.go:134: Deleting namespace: e2e-autoscaled-deployment-rgd-1739345557
    testcase.go:148: Deleting cluster-scoped object: ResourceGraphDefinition/autoscaled-deployment-rgd
--- PASS: TestAutoscaledDeploymentRGD (30.23s)
=== RUN   TestSimpleDeploymentRGD
    simple_deployment_test.go:27: Using namespace: e2e-simple-deployment-rgd-1739345587, from: testdata/../../../../test/e2e/testdata/simple-deployment-rgd
    testcase.go:193: Running step0: Install RGD and verify that the RGD is created
    testcase.go:359:    Applying ResourceGraphDefinition/simple-deployment-rgd[rgd.yaml]
    verify.go:57:       Verifying ResourceGraphDefinition/simple-deployment-rgd[0rgd.yaml]
    verify.go:57:       Verifying CustomResourceDefinition/simpledeployments.kro.run[1instancecrd.yaml]
    testcase.go:193: Running step1: Create instance SimpleDeployment with replicas=2
    testcase.go:359:    Applying SimpleDeployment/test-instance[instance.yaml]
    verify.go:57:       Verifying Deployment/test-instance[deployment.yaml]
    verify.go:57:       Verifying SimpleDeployment/test-instance[instance.yaml]
    testcase.go:193: Running step2: Update instance SimpleDeployment with replicas=3
    testcase.go:359:    Applying SimpleDeployment/test-instance[instance.yaml]
    verify.go:57:       Verifying Deployment/test-instance[deployment.yaml]
    testcase.go:193: Running step3: Update RGD to set the deployment spec to have an empty env array
    testcase.go:359:    Applying ResourceGraphDefinition/simple-deployment-rgd[rgd.yaml]
    verify.go:57:       Verifying ResourceGraphDefinition/simple-deployment-rgd[0rgd.yaml]
    verify.go:57:       Verifying SimpleDeployment/test-instance[1instance.yaml]
    testcase.go:134: Deleting namespace: e2e-simple-deployment-rgd-1739345587
    testcase.go:148: Deleting cluster-scoped object: ResourceGraphDefinition/simple-deployment-rgd
    testcase.go:148: Deleting cluster-scoped object: ResourceGraphDefinition/simple-deployment-rgd
--- PASS: TestSimpleDeploymentRGD (15.21s)
PASS
```

File: pkg/controller/instance/delta/delta.go [hunk 1]
```diff
@@ -113,7 +113,17 @@ func walkCompare(desired, observed interface{}, path string, differences *[]Diff
 		}
 		walkMap(d, e, path, differences)
 
+	case nil:
+		// Special case: treat empty array and nil as equivalent
+		if isEmptyArrayOrNil(desired) && isEmptyArrayOrNil(observed) {
```
Comment (barney-s) [inline pkg/controller/instance/delta/delta.go:123]: Handle for map ?

**Comment by barney-s** [comment]
Few things to try:
1. Identify who is throwing away the empty array ? API server or unstructured.unstructred lib
2. what about CRDs.
3. What about maps{}.

File: pkg/controller/instance/delta/delta.go (hunk 1 above)
Comment (matthchr) [reply pkg/controller/instance/delta/delta.go:123]: Yeah IMO we should probably handle for map too?

File: pkg/controller/instance/delta/delta_test.go [hunk 2]
```diff
@@ -458,3 +458,121 @@ func TestCompare_EmptyMaps(t *testing.T) {
 		})
 	}
 }
+
+func TestCompare_EmptyArraysAndNulls(t *testing.T) {
```
Comment (matthchr) [inline pkg/controller/instance/delta/delta_test.go]: minor: 
```suggestion
func TestCompare_EmptyArraysAndNils(t *testing.T) {
```

File: pkg/controller/instance/delta/delta_test.go [hunk 3]
```diff
@@ -458,3 +458,121 @@ func TestCompare_EmptyMaps(t *testing.T) {
...
+	tests := []struct {
+		name     string
+		desired  *unstructured.Unstructured
+		observed *unstructured.Unstructured
+		wantDiff bool
+	}{
+		{
+			name: "empty array equals null in nested structure",
```
Comment (matthchr) [inline pkg/controller/instance/delta/delta_test.go]: minor:
```suggestion
			name: "empty array equals nil in nested structure",
```

and below

File: pkg/controller/instance/delta/delta_test.go [hunk 4]
```diff
@@ -458,3 +458,121 @@ func TestCompare_EmptyMaps(t *testing.T) {
...
+					},
+				},
+			},
+			observed: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"template": map[string]interface{}{
+							"containers": []interface{}{
+								map[string]interface{}{
+									"env": nil,
```
Comment (matthchr) [inline pkg/controller/instance/delta/delta_test.go:527]: minor: a test that asserts that `map[string]interface{}` with no key at all (no `env`) _does_ differ from one that has an `env` key whose value is nil?

I think that's the expected behavior, right?

**Review by matthchr — COMMENTED** [review]
This looks good to me, but I agree with your comment that we should account for maps too (and add corresponding tests).

File: pkg/controller/instance/delta/delta.go (hunk 1 above)
Comment (barney-s) [reply pkg/controller/instance/delta/delta.go:123]: done.

File: pkg/controller/instance/delta/delta_test.go (hunk 2 above)
Comment (barney-s) [reply pkg/controller/instance/delta/delta_test.go]: done.

File: pkg/controller/instance/delta/delta_test.go (hunk 4 above)
Comment (barney-s) [reply pkg/controller/instance/delta/delta_test.go:527]: yes. added.

**Comment by barney-s** [comment]
> This looks good to me, but I agree with your comment that we should account for maps too (and add corresponding tests).

Done.

File: pkg/controller/instance/delta/delta_test.go [hunk 5]
```diff
@@ -350,6 +350,42 @@ func TestCompare_EmptyMaps(t *testing.T) {
 			},
 			wantDiff: false,
 		},
+		{
```
Comment (matthchr) [inline pkg/controller/instance/delta/delta_test.go:353]: One interesting thing to consider is that we're comparing nil and "omitted" as equal, but we're not comparing omitted and empty as equal, so we have a situation where:

omitted == nil
nil == empty
BUT omitted != empty

Which seems weird?

See these tests:
```
		{
			name: "empty map in desired, no field in observed should diff",
			desired: &unstructured.Unstructured{
				Object: map[string]interface{}{
					"spec": map[string]interface{}{
						"tags": map[string]interface{}{},
					},
				},
			},
			observed: &unstructured.Unstructured{
				Object: map[string]interface{}{
					"spec": map[string]interface{}{},
				},
			},
			wantDiff: true,
		},
		{
			name: "nil map in desired, no field in observed should not diff",
			desired: &unstructured.Unstructured{
				Object: map[string]interface{}{
					"spec": map[string]interface{}{
						"tags": nil,
					},
				},
			},
			observed: &unstructured.Unstructured{
				Object: map[string]interface{}{
					"spec": map[string]interface{}{},
				},
			},
			wantDiff: false,
		},
```

File: pkg/controller/instance/delta/delta_test.go [hunk 6]
```diff
@@ -459,3 +495,121 @@ func TestCompare_EmptyMaps(t *testing.T) {
...
+
+func TestCompare_EmptyArraysAndNil(t *testing.T) {
+	tests := []struct {
+		name     string
+		desired  *unstructured.Unstructured
+		observed *unstructured.Unstructured
+		wantDiff bool
+	}{
+		{
+			name: "empty array equals nil in nested structure",
```
Comment (matthchr) [inline pkg/controller/instance/delta/delta_test.go:507]: Missing tests comparing "omitted" key vs empty or nil?

**Comment by barney-s** [comment]
closing in favor of #561
